src/net/sourceforge/kolmafia/shop/ShopDatabase.java

```diff
@@ -0,0 +1,166 @@
+          continue;
+        }
+
+        String shopId = data[0];
+        String shopType = data[1];
+        String shopName = data[2];
+
+        registerShop(shopId, shopName, Enum.valueOf(SHOP.class, shopType));
+      }
+    } catch (IOException e) {
+      StaticEntity.printStackTrace(e);
+    }
+  }
+
+  public static void registerShop(
+      final String shopId, final String shopName, final CraftingType type) {
+    registerShop(shopId, shopName, SHOP.CONC);
+    shopIdToCraftingType.put(shopId, type);
+    craftingTypeToShopId.put(type, shopId);
+  }
+
+  public static boolean registerShop(
+      final String shopId, final String shopName, final SHOP shopType) {
+    // Return true if this shop is not previously known
+    if (shopIdToShopName.containsKey(shopId)) {
+      // The shopId is already registered. The shopName can differ;
+      // KoL has duplicates, and KoLmafia will internally rename it.
+      //
+      // Visiting a shop will register it as NONE. That's OK.
+      // Some shops are implemented as both NPC and COIN.
+      // Other combos are no good.
+      SHOP existingShopType = shopIdToShopType.get(shopId);
+      if (shopType != existingShopType) {
+        boolean ok =
+            switch (shopType) {
+              case NONE -> true;
+              case NPC -> existingShopType == SHOP.COIN;
+              case COIN -> existingShopType == SHOP.NPC;
+              case CONC -> false;
+            };
+        if (!ok) {
+          String printMe =
+              "Shop id '"
+                  + shopId
+                  + "' of type "
+                  + shopType
+                  + " already registered as "
+                  + existingShopType;
+          RequestLogger.printLine(printMe);
+          RequestLogger.updateSessionLog(printMe);
+        }
+      }
+      return false;
+    }
+
+    shopIdToShopName.put(shopId, shopName);
+    // KoL has duplicate names for shops, but we should not.
+    String existingShopId = shopNameToShopId.get(shopName);
+    if (existingShopId == null) {
+      shopNameToShopId.put(shopName, shopId);
+    } else {
+      String printMe =
+          "Shop name '"
+              + shopName
+              + "' for shop id ' "
+              + shopId
+              + " already used for "
+              + existingShopId;
+      RequestLogger.printLine(printMe);
+      RequestLogger.updateSessionLog(printMe);
+    }
+
+    shopIdToShopType.put(shopId, shopType);
+
+    return true;
+  }
+
+  public static String toData(final String shopId, final String shopName, SHOP shopType) {
+    StringBuilder buf = new StringBuilder();
+    buf.append(shopId);
+    buf.append("\t");
+    buf.append(shopType);
+    buf.append("\t");
+    buf.append(shopName);
+    return buf.toString();
+  }
+
+  public static void writeShopFile() {
+    File output = new File(KoLConstants.DATA_LOCATION, "shops.txt");
+    PrintStream writer = LogStream.openStream(output, true);
+    try {
```

Comment:
if this is `try (writer) {` you don't need the finally, which I think is slightly cleaner.

Reply:
Thanks. I copied from another example, which was presumably written with an older version of Java.